hal/src/common/dmac/transfer.rs

```diff
@@ -0,0 +1,545 @@
+
+        Transfer {
+            buffers: self.buffers,
+            chan,
+            payload: self.payload,
+        }
+    }
+}
+/// These methods are available to a `Transfer` holding a `Ready` channel and a
+/// `BufferPair` holding two arrays of equal type and length
+impl<B, C, const N: usize> Transfer<C, BufferPair<&'static mut [B; N]>>
+where
+    B: 'static + Beat,
+    C: AnyChannel<Status = Ready>,
+{
+    /// Create a new `Transfer` from static array references of the same type
+    /// and length. When two array references are available (instead of slice
+    /// references), it is recommended to use this function over
+    /// [`Transfer::new`](Transfer::new), because it provides compile-time
+    /// checking that the array lengths match. It therefore does not panic, and
+    /// saves some runtime checking of the array lengths.
+    pub fn new_from_arrays(
+        chan: C,
+        source: &'static mut [B; N],
+        destination: &'static mut [B; N],
+        circular: bool,
+    ) -> Self {
+        unsafe { Self::new_unchecked(chan, source, destination, circular) }
+    }
+}
+
+/// These methods are available to a `Transfer` holding a `Busy` channel
+impl<S, D, C, P> Transfer<C, BufferPair<S, D>, P>
+where
+    S: Buffer,
+    D: Buffer<Beat = S::Beat>,
+    C: AnyChannel<Status = Busy>,
+{
+    /// Issue a software trigger request to the corresponding channel.
+    /// Note that is not guaranteed that the trigger request will register,
+    /// if a trigger request is already pending for the channel.
+    #[inline]
+    pub fn software_trigger(&mut self, dmac: &mut DmaController) {
+        // SAFETY: This is safe because we only borrow dmac once.
+        let dmac = unsafe { dmac.dmac_mut() };
+        self.chan.as_mut().software_trigger(dmac);
+    }
+    /// Blocking; Wait for the DMA transfer to complete and release all owned
+    /// resources
+    pub fn wait(self, dmac: &mut DmaController) -> (Channel<ChannelId<C>, Ready>, S, D, P) {
+        // SAFETY: This is safe because we only borrow dmac once.
+        let dmac = unsafe { dmac.dmac_mut() };
+        let chan = self.chan.into().free(dmac);
+
+        // Memory barrier to prevent the compiler/CPU from re-ordering read/write
+        // operations beyond this fence.
+        // (see https://docs.rust-embedded.org/embedonomicon/dma.html#compiler-misoptimizations)
+        atomic::fence(atomic::Ordering::Acquire); // ▼
+
+        (
+            chan,
+            self.buffers.source,
+            self.buffers.destination,
+            self.payload,
+        )
+    }
+
+    /// Non-blocking; Immediately stop the DMA transfer and release all owned
+    /// resources
+    pub fn stop(self, dmac: &mut DmaController) -> (Channel<ChannelId<C>, Ready>, S, D, P) {
+        // SAFETY: This is safe because we only borrow dmac once.
+        let dmac = unsafe { dmac.dmac_mut() };
+        let chan = self.chan.into().stop(dmac);
```

Comment:
Ah, here it is. It didn't show up with a CTRL+F, because the diff was too large and hidden by default.